src/combinatorics/additive/pluennecke_ruzsa.lean

```diff
@@ -0,0 +1,190 @@
+    exact sub_right_injective this }
+end
+
+/-- **Ruzsa's triangle inequality** -/
+lemma card_sub_mul_le_card_add_mul_card_add (A B C : finset α) :
+  (A - C).card * B.card ≤ (A + B).card * (B + C).card :=
+begin
+  rw [←sub_neg_eq_add, ←card_neg B, ←card_neg (B + C), neg_add, ←sub_eq_add_neg],
+  exact card_sub_mul_le_card_sub_mul_card_sub _ _ _,
+end
+
+/-- **Ruzsa's triangle inequality** -/
+lemma card_add_mul_le_card_sub_mul_card_add (A B C : finset α) :
+  (A + C).card * B.card ≤ (A - B).card * (B + C).card :=
+by { rw [←sub_neg_eq_add, ←sub_neg_eq_add B], exact card_sub_mul_le_card_sub_mul_card_sub _ _ _ }
+
+/-- **Ruzsa's triangle inequality** -/
+lemma card_add_mul_le_card_add_mul_card_sub (A B C : finset α) :
+  (A + C).card * B.card ≤ (A + B).card * (B - C).card :=
+by { rw [←sub_neg_eq_add, sub_eq_add_neg B], exact card_sub_mul_le_card_add_mul_card_add _ _ _ }
+
+lemma pluennecke_petridis (C : finset α)
+  (hA : ∀ A' ⊆ A, (A + B).card * A'.card ≤ (A' + B).card * A.card) :
+  (A + B + C).card * A.card ≤ (A + B).card * (A + C).card :=
+begin
+  induction C using finset.induction_on with x C hc ih,
+  { simp },
+  set A' := A ∩ (A + C - {x}) with hA',
+  set C' := insert x C with hC',
+  have h₀ : A' + {x} = (A + {x}) ∩ (A + C),
+  { rw [hA', inter_add_singleton, (is_add_unit_singleton x).sub_add_cancel] },
+  have h₁ : A + B + C' = (A + B + C) ∪ (A + B + {x}) \ (A' + B + {x}),
+  { rw [hC', insert_eq, union_comm, add_union],
+    refine (sup_sdiff_eq_sup _).symm,
+    rw [add_right_comm, add_right_comm A, h₀],
+    exact add_subset_add_right (inter_subset_right _ _) },
+  have h₂ : A' + B + {x} ⊆ A + B + {x} :=
+    add_subset_add_right (add_subset_add_right $ inter_subset_left _ _),
+  have h₃ : (A + B + C').card ≤ (A + B + C).card + (A + B).card - (A' + B).card,
+  { rw h₁,
+    refine (card_union_le _ _).trans_eq _,
+    rw [card_sdiff h₂, ←add_tsub_assoc_of_le (card_le_of_subset h₂), card_add_singleton,
+      card_add_singleton] },
+  refine (mul_le_mul_right' h₃ _).trans _,
+  rw [tsub_mul, add_mul],
+  refine (tsub_le_tsub (add_le_add_right ih _) $ hA _ $ inter_subset_left _ _).trans_eq _,
+  rw [←mul_add, ←mul_tsub, ←hA', insert_eq, add_union, ←card_add_singleton A x,
+    ←card_add_singleton A' x, add_comm (card _), h₀,
+    eq_tsub_of_add_eq (card_union_add_card_inter _ _)],
+end
+
+/-! ### Sum triangle inequality -/
+
+private lemma aux (hA : A.nonempty) (hAB : A ⊆ B)
+  (h : ∀ A' ∈ B.powerset.erase ∅, ((A + C).card : ℚ) / ↑(A.card) ≤ ((A' + C).card) / ↑(A'.card)) :
+  ∀ A' ⊆ A, (A + C).card * A'.card ≤ (A' + C).card * A.card :=
+begin
+  rintro A' hAA',
+  obtain rfl | hA' := A'.eq_empty_or_nonempty,
+  { simp },
+  have hA₀ : (0 : ℚ) < A.card := cast_pos.2 hA.card_pos,
+  have hA₀' : (0 : ℚ) < A'.card := cast_pos.2 hA'.card_pos,
+  exact_mod_cast (div_le_div_iff (hA₀) hA₀').1 (h _ $ mem_erase_of_ne_of_mem hA'.ne_empty $
+    mem_powerset.2 $ hAA'.trans hAB),
+end
+
+/-- **Ruzsa's triangle inequality** -/
```

Comment:
As before, except more because this version is genuinely different from the others!